core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -0,0 +1,216 @@
+    new ZRef[EA, EB, C, B] {
+      def set(c: C): IO[EA, Unit]          = self.set(ca(c))
+      def get: IO[EB, B]                   = self.get
+      def setAsync(c: C): zio.IO[EA, Unit] = self.setAsync(ca(c))
+    }
+
+  /**
+   * Applies a filter to the content of the Ref. If the content of the Ref does not
+   * pass the filter, it will fail in `EB` with `None`
+   */
+  final def filter(f: B => Boolean): ZRef[EA, Option[EB], A, B] =
+    collect { case b if f(b) => b }
+
+  /**
+   * Transforms elements that are retrieved from this Ref with a function.
+   */
+  final def map[C](bc: B => C): ZRef[EA, EB, A, C] =
+    new ZRef[EA, EB, A, C] {
+      def set(a: A): IO[EA, Unit]          = self.set(a)
+      def get: IO[EB, C]                   = self.get.map(bc)
+      def setAsync(a: A): zio.IO[EA, Unit] = self.setAsync(a)
+    }
+
+  /**
+   * Maps from `E` error type to another.
+   */
+  final def mapError[EC](f: EA => EC): ZRef[EC, EB, A, B] = new ZRef[EC, EB, A, B] {
+    def set(a: A): IO[EC, Unit]          = self.set(a).mapError(f)
+    def get: IO[EB, B]                   = self.get
+    def setAsync(a: A): zio.IO[EC, Unit] = self.setAsync(a).mapError(f)
+  }
+
+  /**
+   * Discard the content of the Ref and replace it with `Unit`
+   */
+  final def unit: ZRef[EA, EB, A, Unit] = as(())
+
+  private[zio] def unifyError[E](ea: EA => E, eb: EB => E): ZRef[E, E, A, B] =
+    new ZRef[E, E, A, B] {
+      def set(a: A): IO[E, Unit]          = self.set(a).mapError(ea)
+      def setAsync(a: A): zio.IO[E, Unit] = self.setAsync(a).mapError(ea)
+      def get: IO[E, B]                   = self.get.mapError(eb)
+    }
+
+  /**
+   * Atomically modifies the `Ref` with the specified function. This is not
+   * implemented in terms of `modify` purely for performance reasons.
+   *
+   * @param f function to atomically modify the `Ref`
+   * @return `IO[E, B]` modified value of the `Ref`
+   */
+  def update[E](f: B => A)(implicit ea: EA <:< E, eb: EB <:< E): IO[E, B] = {
+    val unified = unifyError(ea, eb)
+    unified.get.flatMap(b => unified.set(f(b)) *> unified.get).uninterruptible
+  }
+
+  /**
+   * Atomically modifies the `Ref` with the specified function, which computes
+   * a return value for the modification. This is a more powerful version of
+   * `update`.
+   *
+   * @param f function which computes a return value for the modification
+   * @tparam C type of the value of the `Ref` to be modified
+   * @return `IO[E, B]` modified value the `Ref`
+   */
+  def modify[E, C](f: B => (C, A))(implicit ea: EA <:< E, eb: EB <:< E): IO[E, C] = {
+    val unified = unifyError(ea, eb)
+    unified.get.flatMap { b =>
+      val (c, a) = f(b)
+      unified.set(a).as(c)
+    }.uninterruptible
+  }
+
+}
+object ZRef extends Serializable {
+
+  implicit class InvariantSyntax[E, A](ref: ZRef[E, E, A, A]) {
+
+    def updateSome(pf: PartialFunction[A, A]): IO[E, A] =
+      ref.update(pf.orElse { case a => a })
```

Comment:
I think you can write a default version like this in the class itself and move that version in `unsafeCreate`:
```suggestion
      get.flatMap(b => pf.lift(b).fold(ZIO.succeed(b))(a => set(a) *> get).uninterruptible
```